src/common/buffer.cc

```diff
@@ -2150,13 +2150,17 @@ buffer::list buffer::list::static_from_string(string& s) {
 bool buffer::ptr_node::dispose_if_hypercombined(
   buffer::ptr_node* const delete_this)
 {
-  const bool is_hypercombined = static_cast<void*>(delete_this) == \
-    static_cast<void*>(&delete_this->_raw->bptr_storage);
```

Comment:
Hmm, basically we're just calculating the address of `bptr_storage` which should boil down into adding `offsetof(buffer::raw, bptr_storage)` to the value of `_raw` pointer. I doubt any load from `null` is emitted.

UPDATE: there is [the SO answer](https://stackoverflow.com/a/25726076) pointing to a discussion between language's layers.

BTW: `ptr_node` with nullified `_raw` is rather rare / special thing and might be an indicator of other problem. How did it happen in the unittest? Was the `ptr_node` explicitly created without `raw`?

Reply:
> Hmm, basically we're just calculating the address of `bptr_storage` which should boil down into adding `offsetof(buffer::raw, bptr_storage)` to the value of `_raw` pointer. I doubt any load from `null` is emitted.

but this annoys ASan, i will find a way to appease both readability and the sanitizer.

> 
> UPDATE: there is [the SO answer](https://stackoverflow.com/a/25726076) pointing to a discussion between language's layers.

thanks.

> 
> BTW: `ptr_node` with nullified `_raw` is rather rare / special thing and might be an indicator of other problem. How did it happen in the unittest? Was the `ptr_node` explicitly created without `raw`?

it's not rare in crimson, we move `_raw`, `_off` and `_len` from a `ptr` when constructing a `seastar::temporary_buffer<char>` from it. 

https://github.com/ceph/ceph/blob/df10f54f7143fed9b55b693ba4e2723997dfff45/src/common/buffer_seastar.cc#L66-L71

after being moved away, the existing `ptr` is empty. 

this happens in
https://github.com/ceph/ceph/blob/df10f54f7143fed9b55b693ba4e2723997dfff45/src/common/buffer_seastar.cc#L75-L85

where we 

1. we steal the buffers from an existing `buffer::list`. this leaves us a list of empty ptrs.
1. and `clear()` the list, which calls `_buffers.clear_and_dispose()`, it tries to call the deleters of ptrs.

Reply:
i still think it's not **correct**. if `_raw` is null. `offsetof(buffer::raw, bptr_storage) + _raw` is still an invalid address. and we should not consider it when evaluating `is_hypercombined` even if it always evaluates to `false` because `static_cast<void*>(&delete_this->_raw->bptr_storage)` should be a very smaller number in that case.

Reply:
OK, the `clear()` in `list::operator seastar::net::packet()` sanitizes the situation (under HC `ptr_node` shouldn't outlive the `buffer::raw` it's contained within).

> i still think it's not **correct**. if `_raw` is null. `offsetof(buffer::raw, bptr_storage) + _raw` is still an invalid address.

Well, this boils down to a question whether just storing in pointer & comparing a junky address is an UB. How about casting to `std::uintptr_t`? It would cut the dilemma without extra instructions, I hope.

Reply:
> Well, this boils down to a question whether just storing in pointer & comparing a junky address is an UB. How about casting to `std::uintptr_t`? It would cut the dilemma without extra instructions, I hope.

i am not sure this is defined by C++ standard. practically, it is not an undefined behavior. as the lower address spaces are reserved space, .text segment, data segment and bss segment. but even if we cast to `uintptr_t`, the code is still not quite right from two perspectives:

- we rely on the memory layout of a running program arranged by linux kernel here.
- dereference a nullptr annoys the ASan. and it is a false alarm.

Reply:
> - we rely on the memory layout of a running program arranged by linux kernel here.

We're relying on the fact that `ptr_node` can overlap with `buffer::raw` only if these objects are HCed. This is independent from particular OS' memory layout.

> - dereferencing a nullptr (semantically) annoys the ASan. and it is a false alarm.

I see the point. How about:

```cpp
  const bool is_hypercombined = static_cast<std::uintptr_t>(delete_this) == \
    static_cast<std::uintptr_t>(delete_this->_raw) + offsetof(buffer::raw, bptr_storage);
```
?